### ISSUE TITLE
Specifying "outer" type for `rand`

### ISSUE THREAD

**Issue by cscherrer** [cscherrer/MeasureTheory.jl#129]
Great question, I've been fighting with this for a while. Some ideas...

First, we took @devmotion’s suggestion of including an optional float type in the `rand` call. This propagates downward, until eventually influencing the call to `rand(::Type, ::AbstractRNG)`. This is a sort of "inner type".

1. One option would be to have an optional first argument for the "outer type", the constructor for the type to be returned. One issue here is that if the result is nested, say a vector of vectors, it can be awkward to describe in this way. Doesn't seem to compose well.

2. Next, we could base this entirely off the parameters. If you have an `MvNormal` where the mean is a `StaticArray`, maybe you should get another in return. But this is less flexible, and doesn't allow for cases where there's no parameter with the same structure. Just checked again, and see your concern on this as well.

3. Another concern is allocation. So maybe we could have the low-level work entirely in terms of `rand!` Then each user-level `rand` would involve constructing an `undef` and then calling `rand` on the result. I kind of like this one, but for composability we'd also need a way to specify an unboxed array element. This could also maybe work like (1). So far I kind of like this one.

4. Another idea I've played with is to not allocate at all. If samples are sequential, maybe the returned result should be a generator? Then, like (3), the top level could allocate and then traverse the structure and generator together, writing as it goes. This could be ok, but it forces sequential access, and I'm not sure if there's overhead in a traversal like this. 

5. Finally, we could do something similar, but always use a StrideArray. This would make it easy to avoid allocation altogether when we know the size statically, and we could copy it into a different structure at the end, similar to (4). But this would require lots of inlining and `GC.@preserve` s throughout the code, and I think there could be big problems for very large results.

These are all just ideas, but so far (3) seems the best to me.

**Comment by cscherrer** [comment]
Let me know if any of this is unclear or if you have other ideas, and we can discuss :)

**Comment by sethaxen** [comment]
>     1. One option would be to have an optional first argument for the "outer type", the constructor for the type to be returned. One issue here is that if the result is nested, say a vector of vectors, it can be awkward to describe in this way. Doesn't seem to compose well.

Yeah, providing just a type is very limiting, especially because different array types don't all have the same constructor. This is why in manifolds we always require the user to provide a point, whose type we try to respect.

>     2\. Next, we could base this entirely off the parameters. If you have an `MvNormal` where the mean is a `StaticArray`, maybe you should get another in return. But this is less flexible, and doesn't allow for cases where there's no parameter with the same structure. Just checked again, and see your concern on this as well.

That's right, this doesn't really generalize well, and it requires some special-casing, to get around the fact that `similar(x::StaticArray, m, n)` will just be a `Vector`. We basically had to implement our own `similar` machinery to get around this.

>     3\. Another concern is allocation. So maybe we could have the low-level work entirely in terms of `rand!` Then each user-level `rand` would involve constructing an `undef` and then calling `rand` on the result. I kind of like this one, but for composability we'd also need a way to specify an unboxed array element. This could also maybe work like (1). So far I kind of like this one.

I think I prefer this as well. I'm inclined to say we don't _need_ to require the user provide an unboxed array element. They'll get some sensible default if they just call `rand`, and if they want full customizeability of the type, they can allocate their own array and call `rand!`.

I also think this option is the easiest to change from. i.e. if we just implement a bunch of `rand!`s and `default_point(::AbstractMeasure, ::Type)`, with a default `rand` that calls these functions, we know we probably won't want to change `rand!` in the future, just perhaps how `rand` behaves.

For scalar return values, we probably need to use 0-dimensional arrays or `Ref`s; or if we have a way of detecting measures whose points are scalars, we can implement `rand` in that case.

>     4\. Another idea I've played with is to not allocate at all. If samples are sequential, maybe the returned result should be a generator? Then, like (3), the top level could allocate and then traverse the structure and generator together, writing as it goes. This could be ok, but it forces sequential access, and I'm not sure if there's overhead in a traversal like this.

This really won't work for typical ManifoldMeasures cases, where we often need to extensively manipulate a sample from some simpler space: See e.g. https://github.com/JuliaManifolds/ManifoldMeasures.jl/blob/1bf21ca95bc891eac111605c9e3882ce91c18f08/src/stiefel/primitive.jl#L11-L16

>     5\. Finally, we could do something similar, but always use a StrideArray. This would make it easy to avoid allocation altogether when we know the size statically, and we could copy it into a different structure at the end, similar to (4). But this would require lots of inlining and `GC.@preserve` s throughout the code, and I think there could be big problems for very large results.

I'm not really familiar with StrideArrays, but this sounds maybe non-ideal.

**Comment by sethaxen** [comment]
> For scalar return values, we probably need to use 0-dimensional arrays or `Ref`s; or if we have a way of detecting measures whose points are scalars, we can implement `rand` in that case.

Alternatively, we could adopt the convention that our `rand!` signatures are maybe-mutating, that is, they mutate for mutable types and don't for immutable ones. Thus, `rand!(rng, ::T, ::AbstractMeasure) where {T<:Number}` would behave like the current `rand!(rng, ::Type{T}, ::AbstractMeasure) where {T<:Number}`.
